### ISSUE TITLE
Speedup GLMs

### ISSUE THREAD

**Issue by t4c1** [stan-dev/math#1184]
## Description
GLMs can be optimized, especially for small number of parameters. Right now there are multiple instances of unnecessary copying of data and repeating same calculations. Also most loops can be replaced with Eigen expressions that produce vectorized (faster) binary code. There was a previous take on this problem ([pull request 1038](https://github.com/stan-dev/math/pull/1038)).

Further speedups using GPUs are likely possible.

## Example
GLMs are faster.

## Expected Output
GLMs are faster.

#### Current Version:
v2.19.0

**Comment by bob-carpenter** [comment]
Thanks for submitting.  We're trying to coax our issues into being more like technical specs.  For this issue, there needs to be a description of how the GLMs are to be sped up.  I think this issue has already been discussed, so maybe there could be a pointer to that discussion.

**Comment by t4c1** [comment]
I have updated the issue with more details.

**Comment by bob-carpenter** [comment]
Thanks much for adding the details.

**Comment by t4c1** [comment]
I have just completed GPU implementations! Speedups are quite significant:

![GPU speedup k=2000](https://user-images.githubusercontent.com/7499452/56028324-2c93e900-5d18-11e9-9b9d-99e6e42e2374.png)
![GPU speedup k=2](https://user-images.githubusercontent.com/7499452/56028339-2f8ed980-5d18-11e9-8ad9-f48d5951bd76.png)

**Comment by bob-carpenter** [comment]
That's really great.  Is it ready for review?  One of the GPU devs is going to have to take a look.

I take it that K = 2000 and K = 2 are the number of predictors?  

I'm surprised there's any speedup for K = 2 --- is it because things other than the matrix-vector multiply are also being sped up?

Any idea why the bernoulli-logit is so much faster on GPU?  I'd have expected normal_id to speed up the most as it's the one where the matrix-vector multiply should take up most of the cycles.  

Thanks for reporting CPU/GPU types.  This stuff's so hard to evaluate with all the hardware options being so important.

**Comment by t4c1** [comment]
This is waiting for GPU caching and CPU GLM optimizations PRs to be merged. Meanwhile I will probably reorganize the code a bit resulting in another PR that will have to be merged before this one. 

K is the number of attributes of each data point.

Yes, there is a separate kernel for each model that does all the computationally heavy operations.

With a large number of attributes speedups of all four models are comparable. With only a few attributes there are operations, other than matrix-vector multiply that need significant time. While the product is bound by memory transfers, other operations are compute-bound and can be sped up more by using a GPU. In normal-id model there are few other operations so the speedup is lower.

**Comment by SteveBronder** [comment]
> This is waiting for GPU caching

I'll look at that this weekend. Trying to get the async and paper stuff finished up as well

**Comment by rok-cesnovar** [comment]
Closing as this was resolved with #1185, GPU GLMs are part of a new issue #1319
